### ISSUE TITLE
vscode-eslint extension not working on latest version of VSCode, v1.90.0

### ISSUE THREAD

**Issue by D-Maher** [microsoft/vscode-eslint#1856]
## Description
This extension does not appear to be formatting code when setting `"source.fixAll.eslint"` in VSCode's `editor.codeActionsOnSave` settings. When saving a file, the extension hangs indefinitely without much indication in the ESLint Output of what could be going on.

This is an issue I've noticed happening in the last two days or so, and it seems to coincide with updating to the May 2024 release of VSCode, v1.90.0. When downgrading to VSCode v1.89.1, the issue disappears entirely.

I'm not sure if this is an issue with compatibility between the extension and the latest version of VSCode or if this is a bug in VSCode itself, but I'm inclined to think it's the former since other code actions I have configured to run on save (e.g. `"source.organizeImports"`) run just fine on VSCode v1.90.0.

## Additional Information
The extension appears to start up just fine based on what I'm seeing in the ESLint Output:
```
[Info  - 11:18:37 AM] ESLint server is starting.
[Info  - 11:18:38 AM] ESLint server running in node v20.9.0
[Info  - 11:18:38 AM] ESLint server is running.
[Info  - 11:18:39 AM] ESLint library loaded from: /my/obfuscated/path/to/.yarn/sdks/eslint/lib/api.js
```

But when saving a file, I see this pop-up at the bottom right of VSCode that seems to hang there indefinitely:
<img width="377" alt="code_actions_hanging" src="https://github.com/microsoft/vscode-eslint/assets/22512071/b90df3ca-3b43-4530-b818-ae4f21cddeb5">

I've tried hitting 'Cancel' and then running the `ESLint: Restart ESLint Server` command, but then I get the following pop-up:
<img width="382" alt="restart_failed" src="https://github.com/microsoft/vscode-eslint/assets/22512071/efd0e0f7-c651-447a-a2b7-3a4f03ff7415">

And the ESLint output displays the following:
```
[Error - 11:22:35 AM] Restarting client failed
Error: Stopping the server timed out
    at /my/obfuscated/path/to/.vscode/extensions/dbaeumer.vscode-eslint-2.4.4/client/out/extension.js:1:105150
    at b.restart (/my/obfuscated/path/to/.vscode/extensions/dbaeumer.vscode-eslint-2.4.4/client/out/extension.js:1:270303)
```

I'm happy to provide additional information if helpful.

**Comment by cwsault** [comment]
Seeing this as well since the VSCode update; seems to be related to using `eslint-plugin-prettier` & yarn's PNP loading.

I can reproduce it on a fresh project with the following steps, on multiple systems -

1. `yarn create vite some-project-name`
2. choose react & TS; `cd` into project
3. add `"packageManager": "yarn@4.2.2",` to `package.json`
4. `yarn` or `yarn install`
5. `yarn add --dev eslint-plugin-prettier eslint-config-prettier prettier`
6. add `'plugin:prettier/recommended'` to `extends` block of `.eslintrc.cjs`
7. `yarn dlx @yarnpkg/sdks vscode`
8. open project in VSCode; edit & attempt to save a `.ts` or `.tsx` file

You'll also notice if you go to VSCode's Help -> Process Explorer that immediately on open a TS file, the ESLint server will die immediately upon starting (listed as 'defunct') -- probably related to https://github.com/microsoft/vscode-eslint/issues/1855 as well

If `plugin:prettier/recommended` is not included in the `eslintrc` then the problem does not occur. If `yarn` is configured to use `node_modules` installation (`echo nodeLinker: node-modules > `.yarnrc.yml`) then the problem does not occur. Also no issue when using `npm` instead of `yarn`.

**Comment by jrmymbtlr** [comment]
Interesting. I didn't have prettier or prettier extensions installed, but I removed an older .prettierrc and restarted ESLint, and it started working.

I'm almost sure it's just a coincidence.

**Comment by ottobonn** [comment]
We are seeing the same issue after updating to VSCode 1.90, running on Mac OS, using yarn PNP with eslint and prettier and `"editor.codeActionsOnSave": { "source.fixAll": "explicit" }` in our project settings. When saving TypeScript files, the editor waits indefinitely at "Getting Code Actions."

I can also confirm that when restarting the ESLint server, it fails with "stopping the server timed out," but before explicitly trying to restart it, I don't see errors in its output.

**Comment by dbaeumer** [comment]
This is very likely related to https://github.com/yarnpkg/berry/issues/6219.

Can you see if upgrading to yarn 4.3 changes anything.

**Comment by dbaeumer** [comment]
I tired it wit yarn 4.3.0 and I couldn't make the server crash.

**Comment by shermify** [comment]
I tried a bunch of yarn versions including latest build and it doesn't work for me. I'm thinking it might have something to do with the jump to node 20 and changes to the way commonjs/ES modules are resolved.

**Comment by dbaeumer** [comment]
I was able to reproduce this and I am pretty confident that this has nothing to do with the ESLint extension itself. The underlying problem is that prettier is using the syncKit to call sync into a worker which blocks the eslint server. For some reason the worker newer return and therefore the server is hanging here: 

![Image](https://github.com/microsoft/vscode-eslint/assets/1931590/962919d1-e969-4989-95d5-06bee585a6e6)

**Comment by dbaeumer** [comment]
Prettier creates the sync function with:

```
              prettierFormat = require('synckit').createSyncFn(
                require.resolve('./worker'),
              );

```

where `require.resolve('./worker')` resolve to

```
'/workspaces/eslint-plugin-repro/.yarn/__virtual__/eslint-plugin-prettier-virtual-bc1e252fbd/0/cache/eslint-plugin-prettier-npm-5.1.3-496c3b84df-f45d5fc1fc.zip/node_modules/eslint-plugin-prettier/worker.js'
```

It looks like that this makes the worker fail which makes the Atomics.wait never return.

**Comment by dbaeumer** [comment]
This is nothing I can fix in the ESLint extension itself. I opened https://github.com/yarnpkg/berry/issues/6335


Since things seemed to work with Node `18.18.2` you could install that version and use the `eslint.runtime` setting to point to that version.

**Comment by saulrex** [comment]
I was forced to turn off yarn pnp in order to continue working. It's strange why there are so few reviews in this thread.

**Comment by shermify** [comment]
If anyone needs to fix this immediately, you can use this yarn patch on eslint-plugin-prettier. Basically just change the dynamic import in worker.js to a require and point it to prettier/index.cjs

```javascript
diff --git a/eslint-plugin-prettier.js b/eslint-plugin-prettier.js
index 74cd8c0497dadb3a79226bd059845a0fb662a697..cc7b6f7bcb2587d541e042c9e25fd76cf43936f8 100644
--- a/eslint-plugin-prettier.js
+++ b/eslint-plugin-prettier.js
@@ -164,7 +164,7 @@ const eslintPluginPrettier = {
             if (!prettierFormat) {
               // Prettier is expensive to load, so only load it if needed.
               prettierFormat = require('synckit').createSyncFn(
-                require.resolve('./worker'),
+                require.resolve('./worker.js'),
               );
             }
 
diff --git a/worker.js b/worker.js
index 8a8a802ca719a55f4d2ae526eb54e712edfd7455..06269d1c27ca49fd9c7da320318872aa7f40f4f5 100644
--- a/worker.js
+++ b/worker.js
@@ -7,11 +7,10 @@
  */
 
 const { runAsWorker } = require('synckit');
-
+const prettier = require('prettier/index.cjs')
 /**
  * @type {typeof import('prettier')}
  */
-let prettier;
 
 runAsWorker(
   /**
@@ -32,9 +31,6 @@ runAsWorker(
     },
     eslintFileInfoOptions,
   ) => {
-    if (!prettier) {
-      prettier = await import('prettier');
-    }
 
     const prettierRcOptions = usePrettierrc
       ? await prettier.resolveConfig(onDiskFilepath, {
```

**Comment by joonaathaan** [comment]
I've tried the patch, it didn't seem to work sadly.
In the meantime downgrading vsc to april's release seems to be the best option, everything is working properly since it's on node 18.18.2

**Comment by shermify** [comment]
You also need to run yarn unplug eslint-prettier-plugin

**Comment by PavelOparin** [comment]
Hello everyone. For me setting up "eslint.runtime": "node" helped to fix this problem without removing or fixing prettier plugin. Without this settings eslint used node 20.* when I actually have 18.14 and this settings force him to use my version and looks like this solves problem.

**Comment by hanyufoodles** [comment]
By unplugging with `yarn unplug eslint-plugin-prettier` and applying the patch from @shermify, the extension v3.0.10 works again under node v20 and yarn pnp.

**Comment by savolkov** [comment]
This is quite an issue, any plans here?

**Comment by fmal** [comment]
@savolkov this has been fixed with latest release of `synckit` (https://github.com/un-ts/synckit/releases/tag/v0.9.1) which is a dependency of `eslint-plugin-prettier`. Just make sure you have synckit@0.9.1 in your packages. If reinstalling `eslint-plugin-prettier` package doesn't fetch latest `synckit` then you can enforce it through [yarnresolutions](https://yarnpkg.com/configuration/manifest#resolutions) like this:

```json
"resolutions": {
  "synckit": "^0.9.1"
}
```

**Comment by D-Maher** [comment]
Interesting. I now get a different issue after both upgrading `eslint-plugin-prettier` to the latest (v5.2.1) and adding that `"synckit": "^0.9.1"` entry to my `"resolutions"`. Here's my ESLint extension output:

```
[Error - 9:28:25 AM] An unexpected error occurred: undefined Occurred while linting /obfuscated/path/to/MyComponent.tsx:1 Rule: "prettier/prettier".
[Error - 9:28:29 AM] Request textDocument/formatting failed.
  Message: Request textDocument/formatting failed with message: undefined
Occurred while linting /obfuscated/path/to/MyComponent.tsx:1
Rule: "prettier/prettier"
  Code: -32603
```

I've got `"prettier/prettier": "warn"` configured in my ESLint config for `.tsx` files, and removing all Prettier-related configuration in my ESLint config allows this extension to work as expected. Really a bummer since my project is set up to run Prettier through ESLint, but what I'm observing feels more like a `prettier` or `eslint-plugin-prettier` issue at this point.

Just wanted to leave this here in case others run into this too.

<details><summary>Here's my latest environment, FWIW</summary>
<br/>
Operating System: macOS

| Software | Version |
|--------|--------|
| VSCode |`v1.91.1` |
| `vscode-eslint` | `v3.0.10` |
| `eslint` | `v8.46.0` |
| `eslint-plugin-prettier` | `v5.2.1` |
| `prettier` | `v3.3.3` |

Aside from `eslint`, all of the above are latest versions at the time of writing this. It's possible getting `eslint` upgraded to v9 is the key here.
</details>

**Comment by akhmadullin** [comment]
@D-Maher, I have similar problem after adding `"synckit": "^0.9.1"` to  `resolutions`

My project also runs Prettier through ESLint

But text error a little bit shorter, than yours

```
[Error - 09:50:52] An unexpected error occurred: undefined Occurred while linting /path/to/my/file.ts:1 Rule: "prettier/prettier".
``` 

Operating System: macOS

Software | Version
-- | --
VSCode | v1.91.1
vscode-eslint | v3.0.10
eslint | v8.56.0
eslint-plugin-prettier | v5.2.1
eslint-config-prettier | v9.1.0
prettier | v3.3.3

**Comment by fmal** [comment]
> Interesting. I now get a different issue after both upgrading `eslint-plugin-prettier` to the latest (v5.2.1) and adding that `"synckit": "^0.9.1"` entry to my `"resolutions"`. Here's my ESLint extension output:
> 
> ```
> [Error - 9:28:25 AM] An unexpected error occurred: undefined Occurred while linting /obfuscated/path/to/MyComponent.tsx:1 Rule: "prettier/prettier".
> [Error - 9:28:29 AM] Request textDocument/formatting failed.
>   Message: Request textDocument/formatting failed with message: undefined
> Occurred while linting /obfuscated/path/to/MyComponent.tsx:1
> Rule: "prettier/prettier"
>   Code: -32603
> ```
> 
> I've got `"prettier/prettier": "warn"` configured in my ESLint config for `.tsx` files, and removing all Prettier-related configuration in my ESLint config allows this extension to work as expected. Really a bummer since my project is set up to run Prettier through ESLint, but what I'm observing feels more like a `prettier` or `eslint-plugin-prettier` issue at this point.
> 
> Just wanted to leave this here in case others run into this too.
> 
> Here's my latest environment, FWIW

Could you see if this minimal example also fails in your environment: https://github.com/fmal/eslint-prettier-test ? I tested with bunch of projects running eslint@8 with quite complex configs and haven't experienced this issue

**Comment by D-Maher** [comment]
> Could you see if this minimal example also fails in your environment: https://github.com/fmal/eslint-prettier-test ? I updated bunch of projects running eslint@8 to latest `eslint-plugin-prettier` with quite complex configs and haven't experienced the issue you described

That example repo does seem to work just fine in my environment, so it seems like my issue may be more specific to my particular setup. I'm running with a different version of Yarn (v.4.0.2) and using the Yarn SDKs, either of which could be a culprit for me.

**Comment by retrixe** [comment]
> ```
> [Error - 09:50:52] An unexpected error occurred: undefined Occurred while linting /path/to/my/file.ts:1 Rule: "prettier/prettier".
> ```

I was experiencing this issue, and it seems to have been resolved after updating Yarn to 4.4.0 from 4.1.1, then rerunning `yarn dlx @yarnpkg/sdks vscode` to re-create the SDK files. I'm not sure which of the two steps resolved it, but perhaps it resolves the issue on your end too?

**Comment by ben-rogerson** [comment]
> TL;DR: Some combination of upgrading Yarn and upgrading the ESLint-related packages above did the trick. If I had to identify a culprit here, I'm fairly certain it was the eslint-plugin-import (or even the absence of the eslint-import-resolver-typescript package) that was causing this extension to stall out for me.

Thanks for the update - I actually only needed these steps to fix the issue:

1. Update VSCode and restart
2. Update `eslint-plugin-prettier`
3. In vscode, `Developer: Reload Window`

**Comment by RichMatthews** [comment]
This is still an issue, why did it get closed?

**Comment by dbaeumer** [comment]
The issue is very likely related to https://github.com/yarnpkg/berry/issues/6335 which needs to be addressed in yarn.

**Comment by leshasmlesha** [comment]
working in yarn from console, but not working from vscode
